File: edutech-ai/frontend/lib/lessonPlan.ts
// Giáo án điện tử (lesson_plan): kiểu dữ liệu, chuẩn hoá, lựa chọn theo cấp học
// và dựng HTML tự chứa để in / tải về. Mọi chuỗi từ AI đều được escape — file
// xuất ra không có script và bị CSP chặn mọi tài nguyên ngoài.

export interface LessonActivity {
  title: string;
  duration_minutes: number;
  teacher_actions: string[];
  student_actions: string[];
  resources: string[];
}

export interface LessonPlanContent {
  title: string;
  topic: string;
  grade_level: string;
  subject: string;
  class_name: string;
  age_group: string;
  duration_minutes: number;
  objectives: string[];
  materials: string[];
  activities: LessonActivity[];
  assessment: string[];
  differentiation: string[];
  warnings: string[];
}

function str(x: any): string {
  return typeof x === "string" ? x : x == null ? "" : String(x);
}

/** Chấp nhận mảng chuỗi hoặc một chuỗi nhiều dòng; bỏ dòng trống. */
function lines(x: any): string[] {
  const raw = Array.isArray(x) ? x.map(str) : str(x).split(/\r?\n/);
  return raw.map((s) => s.trim()).filter(Boolean);
}

function minutes(x: any): number {
  const n = Math.round(Number(x));
  return Number.isFinite(n) && n > 0 ? n : 0;
}

export function normalizeLessonPlan(content: any): LessonPlanContent {
  const c = content || {};
  const activities: LessonActivity[] = (Array.isArray(c.activities) ? c.activities : []).map(
    (a: any, i: number) => ({
      title: str(a?.title).trim() || `Hoạt động ${i + 1}`,
      duration_minutes: minutes(a?.duration_minutes),
      teacher_actions: lines(a?.teacher_actions),
      student_actions: lines(a?.student_actions),
      resources: lines(a?.resources),
    })
  );
  return {
    title: str(c.title).trim(),
    topic: str(c.topic).trim(),
    grade_level: str(c.grade_level),
    subject: str(c.subject),
    class_name: str(c.class_name),
    age_group: str(c.age_group),
    duration_minutes: minutes(c.duration_minutes),
    objectives: lines(c.objectives),
    materials: lines(c.materials),
    activities,
    assessment: lines(c.assessment),
    differentiation: lines(c.differentiation),
    warnings: lines(c.warnings),
  };
}

// ---------- Lựa chọn theo cấp học ----------

export const GRADE_LABELS: Record<string, string> = {
  mamnon: "Mầm non",
  tieuhoc: "Tiểu học",
  thcs: "THCS",
  thpt: "THPT",
};

export const SUBJECT_LABELS: Record<string, string> = {
  english: "Tiếng Anh",
  mamnon_chung: "Giáo dục mầm non",
  math: "Toán",
  science: "Khoa học tự nhiên",
  literature: "Ngữ văn",
};

/**
 * Nhóm/lớp mầm non (khớp /meta/mindmap của backend) kèm thời lượng một hoạt
 * động học gợi ý theo độ tuổi — giáo viên vẫn sửa được.
 */
export const MAMNON_AGE_GROUPS: { key: string; name_vi: string; minutes: number; range: string }[] = [
  { key: "nha_tre", name_vi: "Nhà trẻ (24–36 tháng)", minutes: 12, range: "10–15 phút" },
  { key: "mg_be", name_vi: "Mẫu giáo bé (3–4 tuổi)", minutes: 18, range: "15–20 phút" },
  { key: "mg_nho", name_vi: "Mẫu giáo nhỡ (4–5 tuổi)", minutes: 25, range: "20–25 phút" },
  { key: "mg_lon", name_vi: "Mẫu giáo lớn (5–6 tuổi)", minutes: 30, range: "25–35 phút" },
];

export function ageGroupName(key: string): string {
  return MAMNON_AGE_GROUPS.find((a) => a.key === key)?.name_vi || key;
}

const CLASS_RANGES: Record<string, [number, number]> = {
  tieuhoc: [1, 5],
  thcs: [6, 9],
  thpt: [10, 12],
};

export function classOptions(grade: string): string[] {
  const r = CLASS_RANGES[grade];
  if (!r) return [];
  const out: string[] = [];
  for (let i = r[0]; i <= r[1]; i++) out.push(`Lớp ${i}`);
  return out;
}

/** Thời lượng một tiết theo cấp: tiểu học 35 phút, THCS/THPT 45 phút. */
export function defaultDuration(grade: string, ageGroup?: string): number {
  if (grade === "mamnon") {
    return MAMNON_AGE_GROUPS.find((a) => a.key === ageGroup)?.minutes || 25;
  }
  return grade === "tieuhoc" ? 35 : 45;
}

export function durationPresets(grade: string): number[] {
  if (grade === "mamnon") return [10, 15, 20, 25, 30, 35];
  if (grade === "tieuhoc") return [35, 70];
  return [45, 90];
}

export const DURATION_MIN = 5;
export const DURATION_MAX = 180;

/** Dòng mô tả lớp/nhóm tuổi + môn, dùng chung cho màn xem, in và trình chiếu. */
export function lessonMetaLine(lp: LessonPlanContent): string {
  return [
    lp.subject ? SUBJECT_LABELS[lp.subject] || lp.subject : "",
    lp.grade_level === "mamnon" && lp.age_group
      ? ageGroupName(lp.age_group)
      : lp.class_name || GRADE_LABELS[lp.grade_level] || lp.grade_level,
    lp.duration_minutes ? `${lp.duration_minutes} phút` : "",
  ]
    .filter(Boolean)
    .join(" · ");
}

export function activitiesTotal(lp: LessonPlanContent): number {
  return lp.activities.reduce((s, a) => s + a.duration_minutes, 0);
}

// ---------- HTML tự chứa ----------

function esc(s: string): string {
  return String(s ?? "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

function list(items: string[], empty = "—"): string {
  if (!items.length) return `<p class="muted">${esc(empty)}</p>`;
  return `<ul>${items.map((i) => `<li>${esc(i)}</li>`).join("")}</ul>`;
}

const HTML_CSS = `
*{box-sizing:border-box}
body{margin:0;padding:32px 20px;background:#f8fafc;color:#1e293b;
font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,sans-serif;line-height:1.55}
main{max-width:960px;margin:0 auto;background:#fff;border:1px solid #e2e8f0;border-radius:16px;padding:36px 40px}
header{border-bottom:2px solid #4f46e5;padding-bottom:16px;margin-bottom:24px}
.eyebrow{margin:0;color:#4f46e5;font-size:12px;font-weight:700;letter-spacing:.12em;text-transform:uppercase}
h1{margin:6px 0 4px;font-size:26px;line-height:1.25}
.meta{margin:0;color:#64748b;font-size:14px}
h2{margin:28px 0 10px;font-size:17px;color:#312e81}
ul{margin:0;padding-left:20px}li{margin:3px 0}
.muted{color:#94a3b8;margin:0}
.grid{display:grid;grid-template-columns:1fr 1fr;gap:24px}
table{width:100%;border-collapse:collapse;font-size:14px}
th,td{border:1px solid #cbd5e1;padding:10px;vertical-align:top;text-align:left}
th{background:#eef2ff;color:#312e81;font-size:13px}
td.act{width:22%;font-weight:600}
.dur{display:inline-block;margin-top:4px;font-weight:500;color:#475569;font-size:12px}
.warn{margin-top:28px;border:1px solid #fcd34d;background:#fffbeb;border-radius:10px;padding:12px 16px;font-size:13px;color:#92400e}
.foot{margin-top:28px;color:#94a3b8;font-size:12px}
@media(max-width:700px){main{padding:22px 18px}.grid{grid-template-columns:1fr}td.act{width:auto}}
@media print{body{background:#fff;padding:0}main{border:0;border-radius:0;padding:0;max-width:none}
tr,li{break-inside:avoid}h2{break-after:avoid}.warn{display:none}}
@page{size:A4;margin:16mm 14mm}
`;

export function buildLessonPlanHtml(lp: LessonPlanContent, fallbackTitle = "Giáo án"): string {
  const title = lp.title || lp.topic || fallbackTitle;
  const rows = lp.activities
    .map(
      (a, i) => `<tr>
<td class="act">${i + 1}. ${esc(a.title)}${
        a.duration_minutes ? `<br><span class="dur">${a.duration_minutes} phút</span>` : ""
      }</td>
<td>${list(a.teacher_actions)}</td>
<td>${list(a.student_actions)}</td>
<td>${list(a.resources)}</td>
</tr>`
    )
    .join("");
  const warnings = lp.warnings.length
    ? `<div class="warn"><strong>Cảnh báo cần rà soát (không in):</strong>${list(lp.warnings)}</div>`
    : "";
  return `<!doctype html>
<html lang="vi">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src 'unsafe-inline'; img-src data:">
<title>${esc(title)}</title>
<style>${HTML_CSS}</style>
</head>
<body>
<main>
<header>
<p class="eyebrow">Kế hoạch bài dạy</p>
<h1>${esc(title)}</h1>
<p class="meta">${esc(lessonMetaLine(lp))}</p>
</header>
<div class="grid">
<section><h2>I. Mục tiêu</h2>${list(lp.objectives)}</section>
<section><h2>II. Chuẩn bị</h2>${list(lp.materials)}</section>
</div>
<h2>III. Tiến trình hoạt động</h2>
${
  lp.activities.length
    ? `<table><thead><tr><th>Hoạt động</th><th>Hoạt động của giáo viên</th><th>Hoạt động của ${
        lp.grade_level === "mamnon" ? "trẻ" : "học sinh"
      }</th><th>Học liệu</th></tr></thead><tbody>${rows}</tbody></table>`
    : `<p class="muted">Chưa có hoạt động.</p>`
}
<div class="grid">
<section><h2>IV. Đánh giá</h2>${list(lp.assessment)}</section>
<section><h2>V. Phân hoá / hỗ trợ</h2>${list(lp.differentiation)}</section>
</div>
${warnings}
<p class="foot">Bản nháp do AI gợi ý — giáo viên rà soát, điều chỉnh cho phù hợp lớp trước khi sử dụng.</p>
</main>
</body>
</html>`;
}

/**
 * In qua iframe ẩn (không mở cửa sổ mới để tránh trình chặn pop-up). HTML đã
 * escape toàn bộ và không có script.
 */
export function printHtml(html: string) {
  const iframe = document.createElement("iframe");
  iframe.setAttribute("aria-hidden", "true");
  iframe.style.position = "fixed";
  iframe.style.right = "0";
  iframe.style.bottom = "0";
  iframe.style.width = "0";
  iframe.style.height = "0";
  iframe.style.border = "0";
  document.body.appendChild(iframe);
  const doc = iframe.contentDocument;
  if (!doc) {
    iframe.remove();
    return;
  }
  doc.open();
  doc.write(html);
  doc.close();
  const run = () => {
    iframe.contentWindow?.focus();
    iframe.contentWindow?.print();
    window.setTimeout(() => iframe.remove(), 1000);
  };
  if (doc.readyState === "complete") window.setTimeout(run, 50);
  else iframe.onload = run;
}

File: edutech-ai/backend/internal/pipelines/lesson_plan.go
package pipelines

import (
	"encoding/json"
	"fmt"
	"strings"
)

// LessonPlanInput is persisted with GradeLevel and Subject injected by the
// backend from the owning project. Values supplied by API clients are ignored.
type LessonPlanInput struct {
	Topic           string `json:"topic"`
	AgeGroup        string `json:"age_group,omitempty"`
	ClassName       string `json:"class_name,omitempty"`
	DurationMinutes int    `json:"duration_minutes,omitempty"`
	Objectives      string `json:"objectives,omitempty"`
	Materials       string `json:"materials,omitempty"`
	Notes           string `json:"notes,omitempty"`
	GradeLevel      string `json:"grade_level"`
	Subject         string `json:"subject"`
}

type LessonPlanActivity struct {
	Title           string `json:"title"`
	DurationMinutes int    `json:"duration_minutes"`
	TeacherActions  string `json:"teacher_actions"`
	StudentActions  string `json:"student_actions"`
	Resources       string `json:"resources"`
}

type lessonPlanModelOutput struct {
	Title           string               `json:"title"`
	Objectives      []string             `json:"objectives"`
	Materials       []string             `json:"materials"`
	Activities      []LessonPlanActivity `json:"activities"`
	Assessment      []string             `json:"assessment"`
	Differentiation []string             `json:"differentiation"`
}

type lessonPlanContent struct {
	Title           string               `json:"title"`
	Topic           string               `json:"topic"`
	GradeLevel      string               `json:"grade_level"`
	Subject         string               `json:"subject"`
	ClassName       string               `json:"class_name"`
	AgeGroup        string               `json:"age_group"`
	DurationMinutes int                  `json:"duration_minutes"`
	Objectives      []string             `json:"objectives"`
	Materials       []string             `json:"materials"`
	Activities      []LessonPlanActivity `json:"activities"`
	Assessment      []string             `json:"assessment"`
	Differentiation []string             `json:"differentiation"`
	Warnings        []string             `json:"warnings"`
}

func lessonPlanSchema() map[string]any {
	stringArray := func() map[string]any {
		return map[string]any{"type": "array", "items": map[string]any{"type": "string"}}
	}
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"title":      map[string]any{"type": "string"},
			"objectives": stringArray(),
			"materials":  stringArray(),
			"activities": map[string]any{
				"type": "array",
				"items": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"title":            map[string]any{"type": "string"},
						"duration_minutes": map[string]any{"type": "integer", "minimum": 1},
						"teacher_actions":  map[string]any{"type": "string"},
						"student_actions":  map[string]any{"type": "string"},
						"resources":        map[string]any{"type": "string"},
					},
					"required": []string{"title", "duration_minutes", "teacher_actions", "student_actions", "resources"},
				},
				"minItems": 1,
			},
			"assessment":      stringArray(),
			"differentiation": stringArray(),
		},
		"required": []string{"title", "objectives", "materials", "activities", "assessment", "differentiation"},
	}
}

const lessonPlanSystemPrompt = `Bạn là chuyên gia thiết kế giáo án điện tử cho giáo viên Việt Nam.
Hãy soạn một giáo án cấp %s, môn/lĩnh vực %s bằng tiếng Việt.
Giáo án phải có mục tiêu quan sát được, hoạt động theo trình tự rõ ràng, đánh giá phù hợp và phương án phân hóa.
Tổng duration_minutes của các hoạt động nên bằng thời lượng tiết học. Không đưa grade_level hoặc subject vào JSON vì hệ thống tự lấy từ project.`

var preschoolAgeNames = map[string]string{
	"nha_tre": "Nhà trẻ 24–36 tháng",
	"mg_be":   "Mẫu giáo bé 3–4 tuổi",
	"mg_nho":  "Mẫu giáo nhỡ 4–5 tuổi",
	"mg_lon":  "Mẫu giáo lớn 5–6 tuổi",
}

func subjectName(key string) string {
	switch key {
	case "english":
		return "Tiếng Anh"
	case "math":
		return "Toán"
	case "science":
		return "Khoa học tự nhiên"
	case "literature":
		return "Ngữ văn"
	case "mamnon_chung":
		return "Giáo dục mầm non"
	default:
		return key
	}
}

func defaultLessonDuration(grade string) int {
	if grade == "mamnon" {
		return 30
	}
	return 45
}

func runLessonPlan(client ollamaChatter, model string, input json.RawMessage) (*Result, error) {
	var in LessonPlanInput
	if err := json.Unmarshal(input, &in); err != nil {
		return nil, fmt.Errorf("input giáo án không hợp lệ: %w", err)
	}
	in.Topic = strings.TrimSpace(in.Topic)
	in.GradeLevel = strings.TrimSpace(in.GradeLevel)
	in.Subject = strings.TrimSpace(in.Subject)
	if in.Topic == "" {
		return nil, fmt.Errorf("cần chủ đề giáo án (topic)")
	}
	if in.GradeLevel == "" || in.Subject == "" {
		return nil, fmt.Errorf("thiếu cấp học hoặc môn học từ project")
	}
	if in.DurationMinutes <= 0 {
		in.DurationMinutes = defaultLessonDuration(in.GradeLevel)
	}

	request, err := json.Marshal(map[string]any{
		"topic":            in.Topic,
		"age_group":        strings.TrimSpace(in.AgeGroup),
		"class_name":       strings.TrimSpace(in.ClassName),
		"duration_minutes": in.DurationMinutes,
		"objectives":       strings.TrimSpace(in.Objectives),
		"materials":        strings.TrimSpace(in.Materials),
		"notes":            strings.TrimSpace(in.Notes),
	})
	if err != nil {
		return nil, err
	}
	system := fmt.Sprintf(lessonPlanSystemPrompt, gradeLevelName(in.GradeLevel), subjectName(in.Subject))
	if in.GradeLevel == "mamnon" {
		ageName := preschoolAgeNames[strings.TrimSpace(in.AgeGroup)]
		if ageName == "" {
			ageName = "Mầm non (chưa chọn nhóm tuổi cụ thể)"
		}
		system += fmt.Sprintf("\nNhóm tuổi: %s. Ưu tiên hoạt động chơi và trải nghiệm, bảo đảm an toàn, phù hợp vận động; không yêu cầu trẻ đọc hoặc viết, đặc biệt với nhà trẻ.", ageName)
	}
	raw, err := client.ChatStructured(model, system, "Yêu cầu giáo án:\n"+string(request), lessonPlanSchema())
	if err != nil {
		return nil, err
	}

	var generated lessonPlanModelOutput
	if err := json.Unmarshal([]byte(raw), &generated); err != nil {
		return nil, fmt.Errorf("model trả JSON giáo án không đúng cấu trúc: %w", err)
	}
	if len(generated.Activities) == 0 {
		return nil, fmt.Errorf("model không sinh hoạt động nào cho giáo án")
	}
	if strings.TrimSpace(generated.Title) == "" {
		generated.Title = "Giáo án: " + in.Topic
	} else {
		generated.Title = strings.TrimSpace(generated.Title)
	}
	if len(generated.Objectives) == 0 {
		generated.Objectives = splitLessonLines(in.Objectives)
	}
	if len(generated.Materials) == 0 {
		generated.Materials = splitLessonLines(in.Materials)
	}
	generated.Objectives = normalizeLessonStrings(generated.Objectives)
	generated.Materials = normalizeLessonStrings(generated.Materials)
	generated.Assessment = normalizeLessonStrings(generated.Assessment)
	generated.Differentiation = normalizeLessonStrings(generated.Differentiation)

	warnings := []string{"Giáo án do AI tạo là bản nháp; giáo viên cần duyệt và điều chỉnh trước khi sử dụng."}
	if len(generated.Objectives) == 0 {
		warnings = append(warnings, "Giáo án chưa có mục tiêu học tập cụ thể")
	}
	if len(generated.Assessment) == 0 {
		warnings = append(warnings, "Giáo án chưa có tiêu chí hoặc hoạt động đánh giá")
	}
	totalActivityMinutes := 0
	for index := range generated.Activities {
		activity := &generated.Activities[index]
		activity.Title = strings.TrimSpace(activity.Title)
		activity.TeacherActions = strings.TrimSpace(activity.TeacherActions)
		activity.StudentActions = strings.TrimSpace(activity.StudentActions)
		activity.Resources = strings.TrimSpace(activity.Resources)
		if strings.TrimSpace(activity.Title) == "" || strings.TrimSpace(activity.TeacherActions) == "" || strings.TrimSpace(activity.StudentActions) == "" {
			return nil, fmt.Errorf("hoạt động %d thiếu tiêu đề hoặc hành động của giáo viên/học sinh", index+1)
		}
		if activity.DurationMinutes <= 0 {
			return nil, fmt.Errorf("hoạt động %d có thời lượng không hợp lệ", index+1)
		}
		totalActivityMinutes += activity.DurationMinutes
	}
	if totalActivityMinutes != in.DurationMinutes {
		warnings = append(warnings, fmt.Sprintf(
			"Tổng thời lượng hoạt động (%d phút) khác thời lượng giáo án (%d phút)",
			totalActivityMinutes, in.DurationMinutes))
	}
	content, err := toContent(lessonPlanContent{
		Title:           generated.Title,
		Topic:           in.Topic,
		GradeLevel:      in.GradeLevel,
		Subject:         in.Subject,
		ClassName:       strings.TrimSpace(in.ClassName),
		AgeGroup:        strings.TrimSpace(in.AgeGroup),
		DurationMinutes: in.DurationMinutes,
		Objectives:      generated.Objectives,
		Materials:       generated.Materials,
		Activities:      generated.Activities,
		Assessment:      generated.Assessment,
		Differentiation: generated.Differentiation,
		Warnings:        []string{},
	})
	if err != nil {
		return nil, err
	}
	return &Result{Title: generated.Title, Content: content, Warnings: warnings}, nil
}

func normalizeLessonStrings(values []string) []string {
	result := make([]string, 0, len(values))
	for _, value := range values {
		value = strings.TrimSpace(value)
		if value != "" {
			result = append(result, value)
		}
	}
	return result
}

func splitLessonLines(value string) []string {
	lines := strings.Split(value, "\n")
	result := make([]string, 0, len(lines))
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			result = append(result, line)
		}
	}
	return result
}
